File: restclient-classic/src/main/java/com/example/demo/PostResourceClient.java
package com.example.demo;

import org.eclipse.microprofile.rest.client.annotation.RegisterProvider;
import org.eclipse.microprofile.rest.client.inject.RegisterRestClient;

import jakarta.ws.rs.*;
import jakarta.ws.rs.core.MediaType;
import java.util.List;
import java.util.concurrent.CompletionStage;

@Path("/posts")
@RegisterRestClient
@RegisterProvider(PostResponseExceptionMapper.class)
public interface PostResourceClient {

    @Path("count")
    @GET
    @Produces(MediaType.TEXT_PLAIN)
    CompletionStage<Long> countAllPosts(@QueryParam("q") String q);

    @GET
    @Produces(MediaType.APPLICATION_JSON)
    CompletionStage<List<com.example.demo.Post>> getAllPosts(
            @QueryParam("q") String q,
            @QueryParam("offset") @DefaultValue("0") int offset,
            @QueryParam("limit") @DefaultValue("10") int limit
    );

    @GET
    @Path("{id}")
    @Produces(MediaType.APPLICATION_JSON)
    CompletionStage<com.example.demo.Post> getPostById(@PathParam("id") String id);
}

File: restclient-kotlin-co/src/main/kotlin/com/example/demo/PostResourceClient.kt
package com.example.demo

import jakarta.ws.rs.*
import jakarta.ws.rs.core.MediaType
import kotlinx.coroutines.flow.Flow
import org.eclipse.microprofile.rest.client.inject.RegisterRestClient

@Path("/posts")
@RegisterRestClient
interface PostResourceClient {

    @GET
    @Produces(MediaType.APPLICATION_JSON)
    suspend fun getAllPosts(
        @QueryParam("q")
        q: String?,
        @QueryParam("offset")
        @DefaultValue("0")
        offset: Int = 0,
        @QueryParam("limit")
        @DefaultValue("10")
        limit: Int = 10
    ): List<Post>

//    does not work when decoding json array to  Flow.
//
//    @GET
//    @Produces(MediaType.APPLICATION_JSON)
//    fun getAllPostsFlow(
//        @QueryParam("q")
//        q: String?,
//        @QueryParam("offset")
//        @DefaultValue("0")
//        offset: Int = 0,
//        @QueryParam("limit")
//        @DefaultValue("10")
//        limit: Int = 10
//    ): Flow<Post>

    @GET
    @Path("count")
    @Produces(MediaType.TEXT_PLAIN)
    suspend fun countAllPosts(
        @QueryParam("q")
        q: String?
    ): Long

}